File: frontend/src/types/transaction.ts
// Enums matching backend
export type Category =
  | 'Equipment'
  | 'Consumables'
  | 'Services'
  | 'CRO_Contract'
  | 'Salaries'
  | 'Overheads'
  | 'Travel'
  | 'Other';

export type GstTreatment =
  | 'CAP'  // Capital
  | 'EXP'  // Expense
  | 'FRE'  // GST Free
  | 'INP'  // Input Taxed
  | 'NTR'  // Not Reported
  | 'MIX'; // Mixed

export type RdRelevance =
  | 'yes'
  | 'partial'
  | 'no';

// Transaction Allocation
export interface TransactionAllocation {
  id: string;
  project_id: string;
  project_name?: string;
  percentage: string;
  amount: string;
}

// Main Transaction interface
export interface Transaction {
  id: string;
  date: string;
  description: string;
  amount: string;
  gst_amount: string | null;
  account_code: string | null;
  reference: string | null;
  supplier_id: string | null;
  supplier_name: string | null;
  category: Category | null;
  gst_treatment: GstTreatment | null;
  rd_relevance: RdRelevance;
  notes: string | null;
  is_reconciled: boolean;
  created_by: string;
  created_at: string;
  updated_at: string;
  allocations: TransactionAllocation[];
}

// Create/Update interfaces
export interface TransactionCreate {
  date: string;
  description: string;
  amount: string;
  gst_amount?: string | null;
  account_code?: string | null;
  reference?: string | null;
  supplier_id?: string | null;
  category?: Category | null;
  gst_treatment?: GstTreatment | null;
  rd_relevance?: RdRelevance;
  notes?: string | null;
}

export interface TransactionUpdate {
  date?: string;
  description?: string;
  amount?: string;
  gst_amount?: string | null;
  account_code?: string | null;
  reference?: string | null;
  supplier_id?: string | null;
  category?: Category | null;
  gst_treatment?: GstTreatment | null;
  rd_relevance?: RdRelevance;
  notes?: string | null;
  project_id?: string | null;
  allocation_percentage?: string;
}

// Filter interface
export interface TransactionFilter {
  date_from?: string;
  date_to?: string;
  category?: Category;
  gst_treatment?: GstTreatment;
  rd_relevance?: RdRelevance;
  supplier_id?: string;
  has_project?: boolean;
  min_amount?: string;
  max_amount?: string;
  search?: string;
}

// Paginated response
export interface PaginatedTransactions {
  items: Transaction[];
  total: number;
  page: number;
  page_size: number;
}

// Summary response
export interface TransactionSummary {
  total_count: number;
  total_amount: string;
  total_gst_amount: string;
  rd_eligible_amount: string;
  by_category: Record<string, string>;
  by_gst_treatment: Record<string, string>;
  by_rd_relevance: Record<string, string>;
}

// Bulk classify request
export interface BulkClassifyRequest {
  transaction_ids: string[];
  category?: Category;
  gst_treatment?: GstTreatment;
  rd_relevance?: RdRelevance;
  project_id?: string;
}

// CSV Import result
export interface CsvRowError {
  row_number: number;
  error: string;
}

export interface CsvImportResult {
  imported_count: number;
  error_count: number;
  errors: CsvRowError[];
}

// Project allocation request
export interface ProjectAllocationRequest {
  project_id: string;
  percentage: string;
}

// Display options for enums
export const CATEGORY_OPTIONS: { value: Category; label: string }[] = [
  { value: 'Equipment', label: 'Equipment' },
  { value: 'Consumables', label: 'Consumables' },
  { value: 'Services', label: 'Services' },
  { value: 'CRO_Contract', label: 'CRO Contract' },
  { value: 'Salaries', label: 'Salaries' },
  { value: 'Overheads', label: 'Overheads' },
  { value: 'Travel', label: 'Travel' },
  { value: 'Other', label: 'Other' },
];

export const GST_TREATMENT_OPTIONS: { value: GstTreatment; label: string }[] = [
  { value: 'CAP', label: 'CAP - Capital' },
  { value: 'EXP', label: 'EXP - Expense' },
  { value: 'FRE', label: 'FRE - GST Free' },
  { value: 'INP', label: 'INP - Input Taxed' },
  { value: 'NTR', label: 'NTR - Not Reported' },
  { value: 'MIX', label: 'MIX - Mixed' },
];

export const RD_RELEVANCE_OPTIONS: { value: RdRelevance; label: string; color: string }[] = [
  { value: 'yes', label: 'Yes - R&D Eligible', color: 'bg-green-100 text-green-700' },
  { value: 'partial', label: 'Partial - R&D Eligible', color: 'bg-yellow-100 text-yellow-700' },
  { value: 'no', label: 'No - Not R&D', color: 'bg-slate-100 text-slate-600' },
];

// Helper functions for display
export function getCategoryLabel(category: Category | null): string {
  if (!category) return '-';
  const option = CATEGORY_OPTIONS.find(opt => opt.value === category);
  return option?.label || category.replace('_', ' ');
}

export function getGstTreatmentLabel(gstTreatment: GstTreatment | null): string {
  if (!gstTreatment) return '-';
  const option = GST_TREATMENT_OPTIONS.find(opt => opt.value === gstTreatment);
  return option?.label || gstTreatment;
}

export function getRdRelevanceLabel(rdRelevance: RdRelevance): string {
  const option = RD_RELEVANCE_OPTIONS.find(opt => opt.value === rdRelevance);
  return option?.label || rdRelevance;
}

export function getRdRelevanceColor(rdRelevance: RdRelevance): string {
  const option = RD_RELEVANCE_OPTIONS.find(opt => opt.value === rdRelevance);
  return option?.color || 'bg-slate-100 text-slate-600';
}

// Format helpers
export function formatCurrency(amount: string | number | null): string {
  if (amount === null || amount === undefined) return '-';
  const num = typeof amount === 'string' ? parseFloat(amount) : amount;
  if (isNaN(num)) return '-';
  return new Intl.NumberFormat('en-AU', {
    style: 'currency',
    currency: 'AUD',
  }).format(num);
}

export function formatDate(dateString: string | null): string {
  if (!dateString) return '-';
  const date = new Date(dateString);
  if (isNaN(date.getTime())) return '-';
  return date.toLocaleDateString('en-AU', {
    day: '2-digit',
    month: '2-digit',
    year: 'numeric',
  });
}

export function formatPercentage(value: string | number | null): string {
  if (value === null || value === undefined) return '-';
  const num = typeof value === 'string' ? parseFloat(value) : value;
  if (isNaN(num)) return '-';
  return `${num.toFixed(2)}%`;
}

File: backend/app/schemas/transaction.py
from datetime import datetime, date
from decimal import Decimal
from typing import Optional, List, Dict, Annotated
from uuid import UUID
from pydantic import BaseModel, Field

from app.models.supplier import Category, GstTreatment
from app.models.transaction import RdRelevance


class TransactionAllocationResponse(BaseModel):
    """Schema for transaction allocation response."""
    id: UUID
    project_id: UUID
    project_name: Optional[str] = None
    percentage: Decimal
    amount: Decimal

    class Config:
        from_attributes = True


class TransactionCreate(BaseModel):
    """Schema for creating a new transaction."""
    date: Annotated[date, Field(description="Transaction date")]
    description: str = Field(..., min_length=1, max_length=500, description="Transaction description")
    amount: Decimal = Field(..., gt=0, description="Transaction amount")
    gst_amount: Optional[Decimal] = Field(None, ge=0, description="GST amount")
    account_code: Optional[str] = Field(None, max_length=50, description="Account code")
    reference: Optional[str] = Field(None, max_length=100, description="Reference number")
    supplier_id: Optional[UUID] = Field(None, description="Supplier ID")
    category: Optional[Category] = Field(None, description="Expense category")
    gst_treatment: Optional[GstTreatment] = Field(None, description="GST treatment")
    rd_relevance: RdRelevance = Field(default=RdRelevance.NO, description="R&D relevance")
    notes: Optional[str] = Field(None, description="Additional notes")


class TransactionUpdate(BaseModel):
    """Schema for updating an existing transaction."""
    date: Annotated[Optional[date], Field(None, description="Transaction date")]
    description: Optional[str] = Field(None, min_length=1, max_length=500, description="Transaction description")
    amount: Optional[Decimal] = Field(None, gt=0, description="Transaction amount")
    gst_amount: Optional[Decimal] = Field(None, ge=0, description="GST amount")
    account_code: Optional[str] = Field(None, max_length=50, description="Account code")
    reference: Optional[str] = Field(None, max_length=100, description="Reference number")
    supplier_id: Optional[UUID] = Field(None, description="Supplier ID")
    category: Optional[Category] = Field(None, description="Expense category")
    gst_treatment: Optional[GstTreatment] = Field(None, description="GST treatment")
    rd_relevance: Optional[RdRelevance] = Field(None, description="R&D relevance")
    notes: Optional[str] = Field(None, description="Additional notes")
    project_id: Optional[UUID] = Field(None, description="Project ID for allocation")
    allocation_percentage: Optional[Decimal] = Field(None, gt=0, le=100, description="Allocation percentage")


class TransactionResponse(BaseModel):
    """Schema for transaction response data."""
    id: UUID
    date: date
    description: str
    amount: Decimal
    gst_amount: Optional[Decimal]
    account_code: Optional[str]
    reference: Optional[str]
    supplier_id: Optional[UUID]
    supplier_name: Optional[str] = None
    category: Optional[Category]
    gst_treatment: Optional[GstTreatment]
    rd_relevance: RdRelevance
    notes: Optional[str]
    is_reconciled: bool
    created_by: UUID
    created_at: datetime
    updated_at: datetime
    allocations: List[TransactionAllocationResponse] = []

    class Config:
        from_attributes = True


class TransactionFilter(BaseModel):
    """Schema for transaction filtering parameters."""
    date_from: Optional[date] = Field(None, description="Filter transactions from this date")
    date_to: Optional[date] = Field(None, description="Filter transactions to this date")
    category: Optional[Category] = Field(None, description="Filter by category")
    gst_treatment: Optional[GstTreatment] = Field(None, description="Filter by GST treatment")
    rd_relevance: Optional[RdRelevance] = Field(None, description="Filter by R&D relevance")
    supplier_id: Optional[UUID] = Field(None, description="Filter by supplier")
    has_project: Optional[bool] = Field(None, description="Filter by project allocation status")
    min_amount: Optional[Decimal] = Field(None, description="Minimum amount filter")
    max_amount: Optional[Decimal] = Field(None, description="Maximum amount filter")
    search: Optional[str] = Field(None, description="Search in description or reference")


class TransactionSummary(BaseModel):
    """Schema for transaction summary/aggregations."""
    total_count: int = Field(..., description="Total number of transactions")
    total_amount: Decimal = Field(..., description="Total amount of all transactions")
    total_gst_amount: Decimal = Field(..., description="Total GST amount")
    rd_eligible_amount: Decimal = Field(..., description="Amount marked as R&D eligible")
    by_category: Dict[str, Decimal] = Field(default_factory=dict, description="Amount by category")
    by_gst_treatment: Dict[str, Decimal] = Field(default_factory=dict, description="Amount by GST treatment")
    by_rd_relevance: Dict[str, Decimal] = Field(default_factory=dict, description="Amount by R&D relevance")


class BulkClassifyRequest(BaseModel):
    """Schema for bulk classification of transactions."""
    transaction_ids: List[UUID] = Field(..., min_length=1, description="List of transaction IDs to update")
    category: Optional[Category] = Field(None, description="New category to set")
    gst_treatment: Optional[GstTreatment] = Field(None, description="New GST treatment to set")
    rd_relevance: Optional[RdRelevance] = Field(None, description="New R&D relevance to set")
    project_id: Optional[UUID] = Field(None, description="Project ID for allocation")


class CsvRowError(BaseModel):
    """Schema for CSV import row-level errors."""
    row_number: int = Field(..., description="Row number in CSV (1-indexed, including header)")
    error: str = Field(..., description="Error message for this row")


class CsvImportResult(BaseModel):
    """Schema for CSV import result."""
    imported_count: int = Field(..., description="Number of successfully imported transactions")
    error_count: int = Field(..., description="Number of rows with errors")
    errors: List[CsvRowError] = Field(default_factory=list, description="List of row-level errors")


class ProjectAllocationRequest(BaseModel):
    """Schema for allocating a transaction to a project."""
    project_id: UUID = Field(..., description="Project ID to allocate to")
    percentage: Decimal = Field(..., gt=0, le=100, description="Allocation percentage")
